core/src/main/scala/org/http4s/rho/bits/ResultResponse.scala

```diff
@@ -1,84 +1,92 @@
 package org.http4s.rho.bits
 
+import cats.data.OptionT
+import cats.{Applicative, Functor, Monad}
+import cats.syntax.functor._
 import org.http4s._
 import org.http4s.rho.Result.BaseResult
-import org.http4s.rho.bits.FailureResponse._
-import org.http4s.rho.bits.ResponseGeneratorInstances.{BadRequest, InternalServerError}
-import fs2.Task
 
 /** Types that represent the result of executing a step of the route */
-sealed trait RouteResult[+T] {
+sealed trait RouteResult[F[_], +T] {
   final def isSuccess: Boolean = this match {
     case SuccessResponse(_) => true
     case _                => false
   }
 
   final def isEmpty: Boolean = this match {
-    case NoMatch => true
-    case _       => false
+    case NoMatch() => true
+    case _        => false
   }
 
-  final def toResponse(implicit ev: T <:< Task[MaybeResponse]): Task[MaybeResponse] = this match {
-      case SuccessResponse(t) => ev(t)
-      case NoMatch            => Pass.now
-      case FailureResponse(r) => r.toResponse
+  final def toResponse(implicit F: Monad[F], ev: T <:< F[Response[F]]): OptionT[F, Response[F]] =
+    this match {
+      case SuccessResponse(t) => OptionT(ev(t).map(Option.apply))
+      case NoMatch()          => OptionT.none[F, Response[F]]
+      case FailureResponse(r) => OptionT(r.toResponse.map(Option.apply))
     }
 }
 
 /** Failure to match a route */
-case object NoMatch extends RouteResult[Nothing]
+case class NoMatch[F[_]]() extends RouteResult[F, Nothing]
```

Comment:
I'm a little queasy of making this a case class since it will be a pretty common result when walking the `PathTree`. A hacky solution (which doesn't fix the part about it being a case class but does fix the root allocation concern) is to make a singleton and cast it to the right type:
```scala
/** Failure to match a route */
case class NoMatch[F[_]] private() extends RouteResult[F, Nothing]

object NoMatch {
  private[this] val singleton = new NoMatch[Nothing]
  // use `get` to not allocate a new `NoMatch`
  def get[F[_]]: NoMatch[F] = singleton.asInstanceOf[NoMatch[F]]
}
```

Reply:
How about putting 

```scala
private val noMatch: RouteResult[F, Nothing] = NoMatch[F]()
```

into `PathTreeOps`? This would allocate it once whenever a `PathTree[F]()` is created, but there would be no need to cast.

Reply:
I'll experiment with @danielkarch 's idea - I would prefer to avoid casting as much as possible

Reply:
ok, commited @danielkarch 's suggestion